### PR TITLE
feat: set groupsync to use default org

### LINKED ISSUE #11938: `setUserGroups` requires 1 org
https://github.com/coder/coder/blob/993c1a67222d328d695665bbccb92b63da78adc3/enterprise/coderd/userauth.go#L27-L33

blocked by https://github.com/coder/coder/issues/11934

### PR DESCRIPTION
**Pull request by Emyrk**
Closes https://github.com/coder/coder/issues/11938

### CODE REVIEW

**Comment by Emyrk** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/coder/coder/12146?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* `main`
  * **#12142** <a href="https://app.graphite.dev/github/pr/coder/coder/12142?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **#12143** <a href="https://app.graphite.dev/github/pr/coder/coder/12143?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **#12145** <a href="https://app.graphite.dev/github/pr/coder/coder/12145?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **#12146** <a href="https://app.graphite.dev/github/pr/coder/coder/12146?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Emyrk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: coderd/userauth.go [hunk 1]
```diff
@@ -1468,8 +1472,36 @@ func (api *API) oauthLogin(r *http.Request, params *oauthLoginParams) ([]*http.C
...
+
+			if !inDefault {
+				return xerrors.Errorf("user %s is not a member of the default organization, cannot assign to groups in the org", user.ID)
+			}
+
 			//nolint:gocritic
-			err := api.Options.SetUserGroups(dbauthz.AsSystemRestricted(ctx), logger, tx, user.ID, filtered, params.CreateMissingGroups)
+			err = api.Options.SetUserGroups(dbauthz.AsSystemRestricted(ctx), logger, tx, user.ID, map[uuid.UUID][]string{
+				defaultOrganization.ID: filtered,
+			}, params.CreateMissingGroups)
```
Comment (Emyrk) [inline coderd/userauth.go:1504]: I do not love how much code this ended up being. I can reduce it by adding another query, but this is all temporary. We need to solve group mapping to orgs, this code is static to only work with the default org.

This PR is required to remove the single org assumption.

File: enterprise/coderd/userauth.go [hunk 2]
```diff
@@ -33,43 +35,49 @@ func (api *API) setUserGroups(ctx context.Context, logger slog.Logger, db databa
...
 
-		if createMissingGroups {
-			// This is the system creating these additional groups, so we use the system restricted context.
-			// nolint:gocritic
-			created, err := tx.InsertMissingGroups(dbauthz.AsSystemRestricted(ctx), database.InsertMissingGroupsParams{
-				OrganizationID: orgs[0].ID,
+		// TODO: This could likely be improved by making these single queries.
+		// 	Either by batching or some other means. This for loop could be really
+		//	inefficient if there are a lot of organizations. There was deployments
+		//	on v1 with >100 orgs.
```
Comment (Emyrk) [inline enterprise/coderd/userauth.go:47]: This is just phase 1 to getting the codebase to accept multi orgs. I do not want to prematurely optimize by coming up with some batching that might be removed.

I think we should move to `pgx` as our pg driver in sqlc to get batching support.

File: coderd/userauth.go [hunk 3]
```diff
@@ -1468,8 +1472,36 @@
...
+				return xerrors.Errorf("get organization memberships: %w", err)
+			}
+
+			inDefault := false
+			for _, membership := range memberships {
+				if membership.OrganizationID == defaultOrganization.ID {
+					inDefault = true
+					break
+				}
+			}
```
Comment (mafredri) [inline coderd/userauth.go:1495]: Tip: https://pkg.go.dev/slices#ContainsFunc

File: coderd/userauth.go [hunk 4]
```diff
@@ -1301,7 +1303,6 @@ func (api *API) oauthLogin(r *http.Request, params *oauthLoginParams) ([]*http.C
 			link database.UserLink
 			err  error
 		)
-
 		user = params.User
 		link = params.Link
```
Comment (mafredri) [inline coderd/userauth.go:1308]: <i class='graphite__hidden'>[Re: line 1347]<br/><br/></i>

Linter is not happy, rebase issue?<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/coder/coder/12146?utm_source=unchanged-line-comment">Graphite</a>.</span>

**Review by mafredri — APPROVED** [review]
LGTM, thanks for adding comments along the way!

**Pull request merged**